lock/postgres.go

```diff
@@ -0,0 +1,109 @@
+package lock
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"time"
+
+	"github.com/sethvargo/go-retry"
+)
+
+// NewPostgresSessionLocker returns a SessionLocker that utilizes PostgreSQL's exclusive
+// session-level advisory lock mechanism.
+//
+// This function creates a SessionLocker that can be used to acquire and release locks for
+// synchronization purposes. The lock acquisition is retried until it is successfully acquired or
+// until the maximum duration is reached. The default lock duration is set to 60 minutes, and the
+// default unlock duration is set to 1 minute.
+//
+// See [SessionLockerOption] for options that can be used to configure the SessionLocker.
+func NewPostgresSessionLocker(opts ...SessionLockerOption) (SessionLocker, error) {
+	cfg := sessionLockerConfig{
+		lockID:         DefaultLockID,
+		lockDuration:   DefaultLockDuration,
+		unlockDuration: DefaultUnlockDuration,
+	}
+	for _, opt := range opts {
+		if err := opt.apply(&cfg); err != nil {
+			return nil, err
+		}
+	}
+	return &postgresSessionLocker{
+		lockID: cfg.lockID,
+		retryLock: retry.WithMaxDuration(
+			cfg.lockDuration,
+			retry.NewConstant(2*time.Second),
+		),
+		retryUnlock: retry.WithMaxDuration(
+			cfg.unlockDuration,
+			retry.NewConstant(2*time.Second),
+		),
+	}, nil
+}
+
+type postgresSessionLocker struct {
+	lockID      int64
+	retryLock   retry.Backoff
+	retryUnlock retry.Backoff
+}
+
+var _ SessionLocker = (*postgresSessionLocker)(nil)
+
+func (l *postgresSessionLocker) SessionLock(ctx context.Context, conn *sql.Conn) error {
+	return retry.Do(ctx, l.retryLock, func(ctx context.Context) error {
+		row := conn.QueryRowContext(ctx, "SELECT pg_try_advisory_lock($1)", l.lockID)
+		var locked bool
+		if err := row.Scan(&locked); err != nil {
+			return err
+		}
+		if locked {
+			// A session-level advisory lock was acquired.
+			return nil
+		}
+		// A session-level advisory lock could not be acquired. This is likely because another
+		// process has already acquired the lock. We will continue retrying until the lock is
+		// acquired or the maximum number of retries is reached.
+		return retry.RetryableError(errors.New("failed to acquire lock"))
+	})
+}
+
+func (l *postgresSessionLocker) SessionUnlock(ctx context.Context, conn *sql.Conn) error {
+	return retry.Do(ctx, l.retryUnlock, func(ctx context.Context) error {
+		var unlocked bool
+		row := conn.QueryRowContext(ctx, "SELECT pg_advisory_unlock($1)", l.lockID)
+		if err := row.Scan(&unlocked); err != nil {
+			return err
+		}
+		if unlocked {
```

Comment:
Could argue this is best effort and goose shouldn't fail if a session cannot be unlocked. Need to think about this scenario a bit more.